.github/workflows/build.yml

```diff
@@ -64,19 +64,134 @@ jobs:
+                }
+              ")
+              if [[ "$has_e2e" == "true" ]]; then
+                # Use the directory path without packages/ prefix (remove trailing slash)
+                package_name="${package_dir%/}"
+                package_name="${package_name#packages/}"
+                packages+=("$package_name")
+              fi
+            fi
+          done
+          # Convert array to JSON array for GitHub Actions matrix
+          if [[ ${#packages[@]} -eq 0 ]]; then
+            echo "No packages with e2e tests found"
+            packages_json="[]"
+          else
+            # Use jq to create a compact single-line JSON array
+            packages_json=$(printf '%s\n' "${packages[@]}" | jq -R '.' | jq -s '.' | jq -c '.')
+          fi
+          echo "packages=$packages_json" >> $GITHUB_OUTPUT
+          echo "Found packages with e2e tests:"
+          echo "$packages_json" | jq -r '.[]' | sed 's/^/  - /'
+
+  test-smoke:
+    needs:
+    - prepare_matrix
+    name: Smoke Tests
+    strategy:
+      matrix:
+        node-version: ${{ fromJSON(needs.prepare_matrix.outputs.versions) }}
+        os:
+        - ubuntu-latest
+    runs-on: ${{ matrix.os }}
+    steps:
+      - uses: actions/checkout@v5
+      - name: Use Node.js ${{ matrix.node-version }}
+        uses: actions/setup-node@v6
+        with:
+          node-version: ${{ matrix.node-version }}
+          cache: 'npm'
+      - name: Install dependencies
+        uses: bahmutov/npm-install@v1
+        with:
+          useRollingCache: true
+          install-command: npm ci --foreground-scripts
+      - name: Build packages
+        run: npm run build
+      - name: Clear V8 code cache (Node.js 24+ compatibility)
+        run: |
+          # Clear V8 code cache to avoid bytecode deserialization errors
+          # Node.js 24 may have cached bytecode that's incompatible
+          if [[ -d "$HOME/.node" ]]; then
+            rm -rf "$HOME/.node"/* 2>/dev/null || true
+          fi
+          if [[ -d "$HOME/.cache/node" ]]; then
+            rm -rf "$HOME/.cache/node"/* 2>/dev/null || true
+          fi
+      - name: Run smoke tests
+        run: npm run test:smoke
+
+  test-unit:
+    needs:
+    - prepare_matrix
+    name: Unit Tests
+    strategy:
+      matrix:
+        node-version: ${{ fromJSON(needs.prepare_matrix.outputs.versions) }}
+        os:
+        - ubuntu-latest
+    runs-on: ${{ matrix.os }}
+    steps:
+      - uses: actions/checkout@v5
+      - name: Use Node.js ${{ matrix.node-version }}
+        uses: actions/setup-node@v6
+        with:
+          node-version: ${{ matrix.node-version }}
+          cache: 'npm'
+      - name: Install dependencies
+        uses: bahmutov/npm-install@v1
+        with:
+          useRollingCache: true
+          install-command: npm ci --foreground-scripts
+      - name: Build packages
+        run: npm run build
+      - name: Clear V8 code cache (Node.js 24+ compatibility)
```

Comment:
Perhaps we could use `&` and `*` to reduce duplication

https://docs.github.com/en/actions/reference/workflows-and-actions/reusing-workflow-configurations#yaml-anchors-and-aliases